tests/main/security-device-cgroups-helper/task.yaml

```diff
@@ -0,0 +1,107 @@
+    # but /dev/kmsg is not
+    udevadm info "/sys/$DEVICES_PATH_MEM_KMSG" | NOMATCH "E: (CURRENT_)?TAGS=.*snap_test-strict-cgroup-helper_sh"
+
+    # the apparmor profile prevents the app from accessing /dev/kmsg workaround
+    # it by switching the profile to complain
+    if aa-status && test -e /var/lib/snapd/apparmor/profiles/snap.test-strict-cgroup-helper.sh; then
+        sed -i -e 's/attach_disconnected,/attach_disconnected,complain,/' \
+            /var/lib/snapd/apparmor/profiles/snap.test-strict-cgroup-helper.sh
+        apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.test-strict-cgroup-helper.sh
+    fi
+
+    snap run test-strict-cgroup-helper.sh -c 'echo hello'
+
+    # explicitly verify that the right cgroup path exists, which is either
+    # /sys/fs/cgroup/devices/snap.test-strict-cgroup-helper.sh or
+    # /sys/fs/bpf/snap/test-strict-cgroup-helper_sh
+    if is_cgroupv2; then
+        test -e /sys/fs/bpf/snap/snap_test-strict-cgroup-helper_sh
+    else
+        test -e /sys/fs/cgroup/devices/snap.test-strict-cgroup-helper.sh
+    fi
+
+    tests.device-cgroup test-strict-cgroup-helper.sh dump | MATCH 'c 1:7 rwm'
+    # /dev/kmsg is not added
+    tests.device-cgroup test-strict-cgroup-helper.sh dump | NOMATCH 'c 1:11 rwm'
+    # and it's not possible to read /dev/kmsg
+    snap run test-strict-cgroup-helper.sh -c 'head -1 /dev/kmsg' 2>&1 | MATCH "Operation not permitted"
+
+    # snap-device-helper is invoked by udev as:
+    # RUN+="$libexecdir/snapd/snap-device-helper $env{ACTION} snap_test-strict-cgroup-helper_sh $devpath $major:$minor"
+
+    "$libexecdir"/snapd/snap-device-helper add snap_test-strict-cgroup-helper_sh "$DEVICES_PATH_MEM_FULL" 1:7
+    "$libexecdir"/snapd/snap-device-helper change snap_test-strict-cgroup-helper_sh "$DEVICES_PATH_MEM_FULL" 1:7
+    # still present
+    tests.device-cgroup test-strict-cgroup-helper.sh dump | MATCH 'c 1:7 rwm'
+
+    # now invoke it for /dev/kmsg, but since the device cgroup is reinitialized
+    # when the snap starts, we have to make the operation async
+    rm -f /var/snap/test-strict-cgroup-helper/common/ready
+    snap run test-strict-cgroup-helper.sh -c 'touch /var/snap/test-strict-cgroup-helper/common/started; until test -e /var/snap/test-strict-cgroup-helper/common/ready; do sleep 1; done; head -1 /dev/kmsg' > run.log 2>&1 &
+    retry -n 5 test -e /var/snap/test-strict-cgroup-helper/common/started
+
+    # device got added
+    "$libexecdir"/snapd/snap-device-helper add snap_test-strict-cgroup-helper_sh "$DEVICES_PATH_MEM_KMSG" 1:11
+    tests.device-cgroup test-strict-cgroup-helper.sh dump | MATCH 'c 1:11 rwm'
+    # or changed
+    "$libexecdir"/snapd/snap-device-helper change snap_test-strict-cgroup-helper_sh "$DEVICES_PATH_MEM_KMSG" 1:11
+    tests.device-cgroup test-strict-cgroup-helper.sh dump | MATCH 'c 1:11 rwm'
+    # and it should be possible to read a line now
+    touch /var/snap/test-strict-cgroup-helper/common/ready
+    wait
```

Comment:
a comment here could explain that this is for the background process we started before